### PR TITLE
Add Analytic Electron-Ion Coupling to CDI

### PR DESCRIPTION
**Pull request by clevelam**
### Background

* This commit adds basic Electron-Ion coupling models to the Common Data Interface (CDI). Thus far, we have only implemented an analytic constant electron-ion coupling model.

### Purpose of Pull Request

* Provide initial Electron-Ion coupling models
* [Electron-Ion coupling #1420](https://rtt.lanl.gov/redmine/issues/1420)

### Description of changes

* Adds EICoupling model to CDI
* This includes the addition of a EICoupling interface class in CDI
* This includes the addition of a derived Analytic_EICoupling interface class that access EICoupling_Model functions
* This includes the addition of an Analytic_EICoupling_Model base class
* This includes the addition of an Analytic_Constant_EICoupling_Model derived class

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

**Comment by clevelam** [comment]
@kgbudge Will you please look over this PR?

**Comment by KineticTheory** [comment]
@clevelam Rerunning the the checks on trinitite didn't fix those issues.  I'll checkout your PR and attempt to run it manually.  Something may have changed on trinitite during this week's DST.

**Comment by KineticTheory** [comment]
@clevelam As I expected, something has changed on trinitite.  I'll consider that a separate issue and won't count the warnings from that machine against this PR.

File: src/cdi/test/tDummyEICoupling.cc [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    message << "The getElectonIonCoupling scalar request"
+            << "request returned a value that is out of spec.";
+    FAILMSG(message.str());
+  }
+
+  std::vector<double> vetemperature(3);
+  std::vector<double> vitemperature(3);
+  vetemperature[0] = 1.0; // Kelvin
+  vetemperature[1] = 3.0; // Kelvin
+  vetemperature[2] = 5.0; // Kelvin
```
Comment (KineticTheory) [inline src/cdi/test/tDummyEICoupling.cc]: Consider using C++ style initializer.

File: src/cdi_analytic/Analytic_EICoupling.cc [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Unpacking constructor.
+ *
+ * This constructor rebuilds and Analytic_EICoupling from a vector<char> that was
+ * created by a call to pack().  It can only rebuild Analytic_Model types
+ * that have been registered in the rtt_cdi_analytic::EICoupling_Models enumeration.
+ */
```
Comment (KineticTheory) [inline src/cdi_analytic/Analytic_EICoupling.cc]: Check for 80 cols max on the above comment.

File: src/cdi_analytic/Analytic_EICoupling.hh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * conforms to the rtt_cdi_analytic::Analytic_EICoupling_Model interface.
+ *
+ * See the member functions for details about the data types and units.
+ *
+ * \example cdi_analytic/test/tstAnalytic_EICoupling.cc
+ *
+ * Example usage of Analytic_EICoupling, Analytic_EICoupling_Model, and their incorporation
+ * into rtt_cdi::CDI.
+ */
+// revision history:
```
Comment (KineticTheory) [inline src/cdi_analytic/Analytic_EICoupling.hh]: Remove 'history' section (git does this for us).  Also check 80 cols max for comments.

File: src/cdi_analytic/Analytic_EICoupling.hh [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+// -----------------
+// 0) original
+//
+//===========================================================================//
+
+class DLL_PUBLIC_cdi_analytic Analytic_EICoupling : public rtt_cdi::EICoupling {
+public:
```
Comment (KineticTheory) [inline src/cdi_analytic/Analytic_EICoupling.hh]: Remove `DLL_PUBLIC_...`.  These are no longer needed except for global static data.

File: src/cdi/EICoupling.hh [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                                        const double density, const double w_e,
+                                        const double w_i) const = 0;
+
+  /*!
+   * \brief EICoupling accessor that returns a vector electron-ion coupling
+   * coefficients.
+   *
+   * \param vetemperature The electron temperature value for which an
+   *     opacity value is being requested (keV).
+   * \param vitemperature The electron temperature value for which an
```
Comment (berselius) [inline src/cdi/EICoupling.hh]: Typo in the docstring here, should be ion (it's obvious, but...)

**Comment by berselius** [comment]
As you mention this can assume a mixture of ion species underneath - it would be good to have a test case that accepts such a mixture.

**Comment by KineticTheory** [comment]
@clevelam This looks to be very close to being done.  Ping me when you think it is ready.

**Comment by KineticTheory** [comment]
Last set of changes looks good. All tests pass.

**Pull request merged**